src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -238,6 +238,71 @@ impl ScalarFunc {
         let week = t.get_time().iso_week().week();
         Ok(Some(i64::from(week)))
     }
+
+    #[inline]
+    pub fn add_datetime_and_duration<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut t: Cow<'a, Time> = try_opt_or!(
+            self.children[0].eval_time(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let d: Cow<'a, MyDuration> = try_opt_or!(
+            self.children[1].eval_duration(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let add = match t
+            .get_time()
+            .checked_add_signed(Duration::nanoseconds(d.to_nanos()))
+        {
+            Some(result) => result,
+            None => return Err(box_err!("parse from duration {} overflows", d)),
+        };
+        let mut res = t.to_mut().clone();
+        res.set_time(add);
```

Comment:
Can we just `t.set_time(add)`?

Reply:
error[E0596]: cannot borrow immutable borrowed content as mutable    
   --> src/coprocessor/dag/expr/builtin_time.rs:296:9
    |
    |         t.set_time(add);
    |         ^ cannot borrow as mutable

```rust
pub fn get_time(&self) -> DateTime<Tz> {...}
pub fn set_time(&mut self, time: DateTime<Tz>) {...}
```

Reply:
Oh, I see, that is a `Cow::Borrowed`, never mind.